doc/source/rllib/single-agent-episode.rst

```diff
@@ -0,0 +1,214 @@
+    Note that these `NDArrays` have an extra batch dim (axis=0), whose length matches the length of the episode stored (here 3).
+
+
+
+Episode.cut() and lookback buffers
+----------------------------------
+
+During sample collection from an RL environment, the :py:class:`~ray.rllib.env.env_runner.EnvRunner` sometimes has to stop
+appending data to an ongoing (non-terminated) :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` to return the
+data collected thus far.
+The `EnvRunner` then calls :py:meth:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode.cut` on
+the :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` object, which returns a
+new episode chunk, with which collection can continue in the next round of sampling.
+
+.. literalinclude:: doc_code/sa_episode.py
+    :language: python
+    :start-after: rllib-sa-episode-04-begin
+    :end-before: rllib-sa-episode-04-end
+
+Note that a "lookback" mechanism exists to allow for connectors to look back into the
+`H` previous timesteps of the cut episode from within the continuation chunk, where `H`
+is a configurable parameter.
+
+.. figure:: images/episodes/sa_episode_cut_and_lookback.svg
+    :width: 800
+    :align: left
+
+The default lookback horizon (`H`) is 1. This means you can - after a `cut()` - still access
+the most recent action (`get_actions(-1)`), the most recent reward (`get_rewards(-1)`),
+and the two most recent observations (`get_observations([-2, -1])`). If you would like to
+be able to access data further in the past, change this setting in your
+:py:class:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig`:
+
+.. testcode::
+    :hide:
+
+    from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
+
+.. testcode::
+
+    config = AlgorithmConfig()
+    # Change the lookback horizon setting, in case your connector (pipelines) need
+    # to access data further in the past.
+    config.env_runners(episode_lookback_horizon=10)
+
+
+Lookback Buffers and getters in more Detail
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following code demonstrates more options available to users of the
+:py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` getter APIs to access
+information further in the past (inside the lookback buffers). Imagine having to write
+a connector piece that has to add the last 5 rewards to the tensor batch used by your model's
+action computing forward pass:
+
+.. literalinclude:: doc_code/sa_episode.py
+    :language: python
+    :start-after: rllib-sa-episode-05-begin
+    :end-before: rllib-sa-episode-05-end
+
+Another useful getter argument (besides `fill`) is the `neg_index_as_lookback` boolean argument.
+If set to True, negative indices are not interpreted as "from the end", but as
+"into the lookback buffer". This allows you to loop over a range of global timesteps
```

Comment:
This is a tiny tiny bit ambiguous when the "lookback" is actually not into the "lookback buffer" but just backwards from the considered global timestep. So, precisely it would be "from the global timestep (if needed into the lookback buffer)", wouldn't it?

Reply:
I think this is fine.

Note that we do not support indexing by global (actual gym episode) timestep yet. It's probably a usecase that's never really needed.

So if you have an episode chunk from global timestep 100 to global timestep 200, as a user, you would normally just loop through it (from `local_index` 0 to `local_index` 100) and then - at each `local_index` - say: I want to do 4x-framestacking from `local_index - 3` to `local_index + 1`. This means that if local_index is 0 or 1 or 2, you'd have a negative start index and for these few cases you want to tell the getter that these negative indices indeed mean "go into the lookback buffer" instead of "count from the end".

In other words: If you want to index into the lookback buffer, you have to provide a negative number, b/c `index=0` is always the first item after(!) the lookback buffer.